src/vm/types/programjson.zig

```diff
@@ -0,0 +1,654 @@
+        if (!std.mem.eql(u8, PRIME_STR, self.prime))
+            return ProgramError.PrimeDiffers;
+
+        // Finds the program counter of the specified entrypoint identifier.
+        // Returns an error if the identifier is not found.
+        const entrypoint_pc = if (entrypoint) |e| blk: {
+            const key = try std.mem.concat(
+                allocator,
+                u8,
+                &[_][]const u8{ "__main__.", e.* },
+            );
+            defer allocator.free(key);
+
+            if (self.identifiers.map.get(key)) |entrypoint_identifier| {
+                break :blk entrypoint_identifier.pc;
+            } else {
+                return ProgramError.EntrypointNotFound;
+            }
+        } else null;
+
+        const start = if (self.identifiers.map.get("__main__.__start__")) |identifier| identifier.pc else null;
+        const end = if (self.identifiers.map.get("__main__.__end__")) |identifier| identifier.pc else null;
+
+        var constants = std.StringHashMap(Felt252).init(allocator);
+        errdefer constants.deinit();
+        // Populates the constants hashmap with identifier keys and associated constant values.
+        for (self.identifiers.map.keys(), self.identifiers.map.values()) |key, value| {
+            if (value.type) |_| {
+                try constants.put(
+                    key,
+                    if (value.value) |v| Felt252.fromInteger(@intCast(v)) else return ProgramError.ConstWithoutValue,
+                );
+            }
+        }
+
+        // Collects attributes named "error_message" and adds them to error_message_attributes.
+        var error_message_attributes = std.ArrayList(Attribute).init(allocator);
+        errdefer error_message_attributes.deinit();
+        for (0..self.attributes.len) |i| {
+            const name = self.attributes[i].name;
+            if (std.mem.eql(u8, name, "error_message")) {
+                try error_message_attributes.append(self.attributes[i]);
+            }
+        }
+
+        var builtins = std.ArrayList(BuiltinName).init(allocator);
+        errdefer builtins.deinit();
+        // Collects built-in names and adds them to the builtins list.
+        for (0..self.attributes.len) |i| {
+            try builtins.append(std.meta.stringToEnum(BuiltinName, self.builtins[i]) orelse return ProgramError.BuiltinNotInLayout);
+        }
+
+        var instruction_locations = std.StringHashMap(InstructionLocation).init(allocator);
+        errdefer instruction_locations.deinit();
+        // Populates the instruction_locations hashmap with debug information related to instruction locations.
+        for (self.debug_info.instruction_locations.map.keys(), self.debug_info.instruction_locations.map.values()) |key, value| {
+            try instruction_locations.put(key, value);
+        }
+
+        var identifiers = std.StringHashMap(Identifier).init(allocator);
```

Comment:
the underlying impl of the json hashmap is `StringArrayHashMapUnmanaged`: StringArrayHashMapUnmanaged

so I am curious if we can clone here
https://github.com/ziglang/zig/blob/master/lib/std/array_hash_map.zig#L385

Reply:
So yes I looked a little in fact we could do that but:
- We would have to go from a simple hashmap to an arrayHashMap and I don't think we necessarily need it.
- We would have to have an unmanaged arrayHashMap, which is a little more complicated to handle afterwards.

So I told myself that the simplest solution was to do it like this but I agree that it's a bit repetitive to go through all the keys and values again.

What do you think?